src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,583 @@
+
+        // Limiting concurrent requests to limit the amount of files open at a time on Mac OSX
+        // the default is 256 which is easy to hit if we don't limit concurrency
+        private readonly static SemaphoreSlim _throttle =
+            RuntimeEnvironmentHelper.IsMacOSX
+                ? new SemaphoreSlim(ConcurrencyLimit, ConcurrencyLimit)
+                : null;
+
+        public HttpSource(PackageSource source, Func<Task<HttpHandlerResource>> messageHandlerFactory)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            if (messageHandlerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(messageHandlerFactory));
+            }
+
+            _packageSource = source;
+            _baseUri = new Uri(source.Source);
+            _messageHandlerFactory = messageHandlerFactory;
+        }
+
+        internal Task<HttpSourceResult> GetAsync(string uri,
+            string cacheKey,
+            HttpSourceCacheContext context,
+            ILogger log,
+            CancellationToken cancellationToken)
+        {
+            return GetAsync(uri, cacheKey, context, log, ignoreNotFounds: false, cancellationToken: cancellationToken);
+        }
+
+        /// <summary>
+        /// Caching Get request.
+        /// </summary>
+        internal async Task<HttpSourceResult> GetAsync(string uri,
+            string cacheKey,
+            HttpSourceCacheContext cacheContext,
+            ILogger log,
+            bool ignoreNotFounds,
+            CancellationToken cancellationToken)
+        {
+            var sw = new Stopwatch();
+            sw.Start();
+
+            var result = await TryCache(uri, cacheKey, cacheContext, cancellationToken);
+            if (result.Stream != null)
+            {
+                log.LogInformation(string.Format(CultureInfo.InvariantCulture, "  {0} {1}", "CACHE", uri));
+                return result;
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Get, uri);
+            log.LogInformation(string.Format(CultureInfo.InvariantCulture, "  {0} {1}.", "GET", uri));
+
+            // Read the response headers before reading the entire stream to avoid timeouts from large packages.
+            Func<Task<HttpResponseMessage>> throttleRequest = () => SendWithCredentialSupportAsync(
+                    request,
+                    HttpCompletionOption.ResponseHeadersRead,
+                    cancellationToken);
+
+            using (var response = await GetThrottled(throttleRequest, log))
+            {
+                if (ignoreNotFounds && response.StatusCode == HttpStatusCode.NotFound)
+                {
+                    log.LogInformation(string.Format(CultureInfo.InvariantCulture,
+                        "  {1} {0} {2}ms", uri, response.StatusCode.ToString(), sw.ElapsedMilliseconds.ToString()));
```

Comment:
Is this string going to make sense when localized, and will the authors understand what it is displaying since the only text is ms?